Update ES span content to include db.statement

rmfitzpatrick opened the pull request:
These changes introduce the `SIGNALFX_INSTRUMENTATION_ELASTICSEARCH_TAG_QUERIES` environment variable to allow tagging `db.statement` from query post data.  Also updates the operation name taken from the source request parameters to be more informative.

johnbley commented on test/Datadog.Trace.ClrProfiler.IntegrationTests/Elasticsearch5Tests.cs:
```diff
@@ -122,18 +138,45 @@ public void SubmitsTraces(string packageVersion)
                                  .OrderBy(s => s.Start)
                                  .ToList();
 
+                var statementNames = new List<string>
```
You presumably got this list from someplace.  A comment with a link to the URL / class name / etc. helps future maintainers greatly here as it becomes that much easier to check "hey, is the list up to date with new features?"

rmfitzpatrick replied:
My initial method was grepping for the password data member, and this is not ideal.  Updated to sanitize to strip usernames and passwords.

rmfitzpatrick replied:
Realized I was likely misunderstanding your point.  These operation names come from the tested sample programs, so if something that's reported is missing upon integration test modification the test will fail.

johnbley commented on src/Datadog.Trace.ClrProfiler.Managed/Integrations/ElasticsearchNetCommon.cs:
```diff
@@ -68,5 +71,100 @@ public static Scope CreateScope(Tracer tracer, string integrationName, object pi
+        public static void SetDbStatement(Span span, object writtenBytes)
+        {
+            string data = System.Text.Encoding.UTF8.GetString((byte[])writtenBytes);
+            string statement = data.Length > 1024 ? data.Substring(0, 1024) : data;
```
For performance overhead it would be much better to apply this cap/limit when the stream reads this data from the source rather than truncating after the fact.

rmfitzpatrick replied:
Done, though getting the written bytes via the postData method does require complete serialization.

pjanotti commented on src/Datadog.Trace.ClrProfiler.Managed/Integrations/ElasticsearchNetCommon.cs:
```diff
@@ -68,5 +71,100 @@ public static Scope CreateScope(Tracer tracer, string integrationName, object pi
+
+        public static void SetDbStatement(Span span, object writtenBytes)
+        {
+            string data = System.Text.Encoding.UTF8.GetString((byte[])writtenBytes);
```
It moved by the Q is still applicable.

rmfitzpatrick replied:
Updated

rmfitzpatrick commented on src/Datadog.Trace.ClrProfiler.Managed/Integrations/ElasticsearchNetCommon.cs:
```diff
@@ -68,5 +71,100 @@ public static Scope CreateScope(Tracer tracer, string integrationName, object pi
+            return true;
+        }
+
+        public static MethodInfo GetWriteMethodInfo(string methodName, object requestData, object postData, out object connectionSettings)
```
It's my understanding that this isn't possible without placing a dependency on the Elasticsearch.Net library, which the reflection-driven helpers and method builders in instrumentations make unnecessary and likely undesired.

rmfitzpatrick commented on src/Datadog.Trace.ClrProfiler.Managed/Integrations/ElasticsearchNetCommon.cs:
```diff
@@ -68,5 +71,100 @@ public static Scope CreateScope(Tracer tracer, string integrationName, object pi
+            if (writtenBytes == null)
+            {
+                object connectionSettings;
+                var methodInfo = GetWriteMethodInfo("WriteAsync", requestData, postData, out connectionSettings);
```
Shared w/ previous response

pjanotti commented on src/Datadog.Trace.ClrProfiler.Managed/Integrations/ElasticsearchNetCommon.cs:
```diff
@@ -68,5 +75,122 @@ public static Scope CreateScope(Tracer tracer, string integrationName, object pi
             return scope;
         }
+
+        public static bool AttemptWrittenBytes(Span span, object requestData, out object postData, out byte[] writtenBytes)
```
nit: I think these methods deserve comments on what they are doing.

rmfitzpatrick commented:
@pjanotti, thanks for your feedback.  I've updated with xml comments and refactored slightly.  Would appreciate another pass to see if this clears up the code when you get a chance.
